### PR TITLE
[Merged by Bors] - feat: roots in an algebra

### PR DESCRIPTION
**Pull request by astrainfinita**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/FieldTheory/SplittingField/Construction.lean [hunk 1]
```diff
@@ -195,13 +195,13 @@ theorem adjoin_rootSet (n : ℕ) :
     ∀ {K : Type u} [Field K],
       ∀ (f : K[X]) (_hfn : f.natDegree = n),
         Algebra.adjoin K
-            (↑(f.map <| algebraMap K <| SplittingFieldAux n f).roots.toFinset :
-              Set (SplittingFieldAux n f)) = ⊤)
+          ((f.aroots (SplittingFieldAux n f)).toFinset : Set (SplittingFieldAux n f)) = ⊤)
```
Comment (tb65536) [inline Mathlib/FieldTheory/SplittingField/Construction.lean:198]: Is this actually just `rootSet`?

**Review by tb65536 — APPROVED** [review]
This looks great! Thanks!

File: Mathlib/FieldTheory/Separable.lean [hunk 2]
```diff
@@ -575,7 +575,7 @@ variable [Algebra K S] [Algebra K L]
 theorem AlgHom.card_of_powerBasis (pb : PowerBasis K S) (h_sep : (minpoly K pb.gen).Separable)
     (h_splits : (minpoly K pb.gen).Splits (algebraMap K L)) :
     @Fintype.card (S →ₐ[K] L) (PowerBasis.AlgHom.fintype pb) = pb.dim := by
-  let s := ((minpoly K pb.gen).map (algebraMap K L)).roots.toFinset
+  let s := ((minpoly K pb.gen).aroots L).toFinset
```
Comment (Ruben-VandeVelde) [inline Mathlib/FieldTheory/Separable.lean:578]: This is also `rootSet`

File: Mathlib/Data/Polynomial/RingDivision.lean [hunk 3]
```diff
@@ -914,25 +914,43 @@ theorem funext [Infinite R] {p q : R[X]} (ext : ∀ r : R, p.eval r = q.eval r)
...
+    p.rootSet S = (p.aroots S).toFinset := by
   rw [rootSet]
   convert rfl
 #align polynomial.root_set_def Polynomial.rootSet_def
 
 @[simp]
 theorem rootSet_C [CommRing S] [IsDomain S] [Algebra T S] (a : T) : (C a).rootSet S = ∅ := by
   classical
-  rw [rootSet_def, map_C, roots_C, Multiset.toFinset_zero, Finset.coe_empty]
+  rw [rootSet_def, aroots_def, map_C, roots_C, Multiset.toFinset_zero, Finset.coe_empty]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Polynomial/RingDivision.lean:953]: Should we prove this in terms of a new `aroots_C` lemma?

File: Mathlib/FieldTheory/Galois.lean [hunk 4]
```diff
@@ -396,7 +396,7 @@ theorem of_separable_splitting_field [sp : p.IsSplittingField F E] (hp : p.Separ
     IsGalois F E := by
   haveI hFE : FiniteDimensional F E := Polynomial.IsSplittingField.finiteDimensional E p
   letI := Classical.decEq E
-  let s := (p.map (algebraMap F E)).roots.toFinset
+  let s := (p.aroots E).toFinset
```
Comment (Ruben-VandeVelde) [inline Mathlib/FieldTheory/Galois.lean:399]: `rootSet`

File: Mathlib/FieldTheory/Separable.lean (hunk 2 above)
Comment (negiizhao) [reply Mathlib/FieldTheory/Separable.lean:578]: `Fintype.card_of_subtype` requires a `Finset`

Comment (negiizhao) [reply Mathlib/FieldTheory/Separable.lean:578]: Ah but `s` can be inferred automatically.

File: Mathlib/Data/Polynomial/RingDivision.lean [hunk 5]
```diff
@@ -914,25 +914,107 @@ theorem funext [Infinite R] {p q : R[X]} (ext : ∀ r : R, p.eval r = q.eval r)
 
 variable [CommRing T]
 
-/-- The set of distinct roots of `p` in `E`.
+/-- `aroots p A` gives the multiset of the roots of `p` in an algebra `A`,
+including their multiplicities. -/
```
Comment (ocfnash) [inline Mathlib/Data/Polynomial/RingDivision.lean:918]:
```suggestion
/-- Given a polynomial `p` with coefficients in a ring `T` and a `T`-algebra `S`, `aroots p S` is
the multiset of roots of `p` regarded as a polynomial over `S`. -/
```

File: Mathlib/Data/Polynomial/RingDivision.lean [hunk 6]
```diff
@@ -914,25 +914,107 @@ theorem funext [Infinite R] {p q : R[X]} (ext : ∀ r : R, p.eval r = q.eval r)
...
+    [NoZeroSMulDivisors T S] {p : T[X]} {a : S} : a ∈ p.aroots S ↔ p ≠ 0 ∧ aeval a p = 0 := by
+  rw [mem_aroots', Polynomial.map_ne_zero_iff]
+  exact NoZeroSMulDivisors.algebraMap_injective T S
+
+theorem aroots_mul [CommRing S] [IsDomain S] [Algebra T S]
+    [NoZeroSMulDivisors T S] {p q : T[X]} (hpq : p * q ≠ 0) :
+    (p * q).aroots S = p.aroots S + q.aroots S := by
+  rw [aroots_def, Polynomial.map_mul, roots_mul]
+  rwa [← Polynomial.map_mul, Polynomial.map_ne_zero_iff]
+  exact NoZeroSMulDivisors.algebraMap_injective T S
```
Comment (ocfnash) [inline Mathlib/Data/Polynomial/RingDivision.lean:940]: Spawning goals by calling rewrite with missing hypotheses makes proof hard to follow and maintain. It's usually better to establish what hypotheses you need first and then pass them in. For example, here we could do:
```suggestion
  suffices : map (algebraMap T S) p * map (algebraMap T S) q ≠ 0
  · rw [aroots_def, Polynomial.map_mul, roots_mul this]
  rwa [← Polynomial.map_mul, Polynomial.map_ne_zero_iff
    (NoZeroSMulDivisors.algebraMap_injective T S)]
```

File: Mathlib/Data/Polynomial/RingDivision.lean [hunk 7]
```diff
@@ -972,14 +1054,13 @@ theorem bUnion_roots_finite {R S : Type*} [Semiring R] [CommRing S] [IsDomain S]
...
-    aeval_def]
+  rw [rootSet_def, Finset.mem_coe, mem_toFinset, mem_aroots']
 #align polynomial.mem_root_set' Polynomial.mem_rootSet'
 
 theorem mem_rootSet {p : T[X]} {S : Type*} [CommRing S] [IsDomain S] [Algebra T S]
     [NoZeroSMulDivisors T S] {a : S} : a ∈ p.rootSet S ↔ p ≠ 0 ∧ aeval a p = 0 := by
-  rw [mem_rootSet',
-    (map_injective _ (NoZeroSMulDivisors.algebraMap_injective T S)).ne_iff' (Polynomial.map_zero _)]
+  rw [mem_rootSet', Polynomial.map_ne_zero_iff]
+  exact NoZeroSMulDivisors.algebraMap_injective T S
```
Comment (ocfnash) [inline Mathlib/Data/Polynomial/RingDivision.lean:1063]:
```suggestion
  rw [mem_rootSet', Polynomial.map_ne_zero_iff (NoZeroSMulDivisors.algebraMap_injective T S)]
```

**Review by ocfnash — APPROVED** [review]
Thanks; please apply the suggestions and then feel free to merge!

bors d+
